feat: deprecate sp-menu in PickerBase derived classes

adixon-adobe opened the pull request:
## Description

This is an update to #1198 with the functional approach. It should also address some of the test failures there.

I'll leave as a draft until I update all the related docs/tests.

## Related Issue

#1235 

## Motivation and Context

Simpler APIs are generally better.

## How Has This Been Tested?

Just some basic doc page testing so far.

## Screenshots (if appropriate):

## Types of changes

- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [] I have added tests to cover my changes.
- [] All new and existing tests passed.

adixon-adobe commented on packages/picker/src/Picker.ts:
```diff
@@ -256,32 +251,27 @@ export class PickerBase extends SizedMixin(Focusable) {
             placement: this.placement,
             receivesFocus: 'auto',
         });
+        await this.manageSelection();
```
Because we're now reparenting into the `sp-menu`, we need to manage selection for the menu to properly update. @Westbrook let me know if this approach makes sense. I'm not totally confident I'm covering everything yet.

adixon-adobe replied:
Yeah -- we'll want to do this in the next iteration. I can already see how #1189 will simplify a lot of the internal logic here.

adixon-adobe commented:
@Westbrook there's still a few SplitButton failures, and I now see that the markup is a little different for it.

Is there a reason SplitButton needs the client to create the popover as well?

Westbrook reviewed (changes requested):
While you're here can you confirm that the slot documentation for each of these element is correct based on this change (not to say it was 100% correct before). For instance, this was wrong before and is more wrong now: https://github.com/adobe/spectrum-web-components/blob/main/packages/split-button/src/SplitButton.ts#L44 Action Menu looks like it has the same twice out of date listing while Picker seems like it just needs to be specifically updated for this change.

Westbrook commented:
Also, the SplitButton test failures all originate in how that component manages it's items a little different: https://github.com/adobe/spectrum-web-components/blob/main/packages/split-button/src/SplitButton.ts#L180

adixon-adobe commented on packages/picker/src/Picker.ts:
```diff
@@ -256,32 +252,27 @@ export class PickerBase extends SizedMixin(Focusable) {
             !this.optionsMenu ||
-            this.optionsMenu.children.length === 0
+            !this.reparentableChildren ||
+            this.reparentableChildren.length === 0
```
ahhh -- yeah. Not used to typescript yet :)

adixon-adobe commented on packages/picker/src/Picker.ts:
```diff
@@ -256,32 +252,27 @@ export class PickerBase extends SizedMixin(Focusable) {
         if (
             !this.popover ||
```
yes

adixon-adobe commented on packages/picker/src/Picker.ts:
```diff
@@ -354,19 +345,31 @@ export class PickerBase extends SizedMixin(Focusable) {
         super.firstUpdated(changedProperties);
 
-        this.optionsMenu = this.querySelector('sp-menu') as Menu;
+        this.optionsMenu = this.shadowRoot.querySelector('sp-menu') as Menu;
```
oh hey, we can. Nice!

adixon-adobe replied:
oh hey, we can't quite (this still gets reparented by the popover, so that refactor leads to null results when the menu is open)

adixon-adobe replied:
Double-check my changes here. Is it okay to have the `optionsMenu!` if it gets initialized in `firstUpdated`?

Westbrook replied:
In theory `@query('sp-menu', true)` would cache this, but if the timing still doesn't work out right, no problem.

We "know" the options menu will be set everywhere it's used, so we can still use the `!` in my oppinion.

adixon-adobe commented on packages/shared/src/reparent-children.ts:
```diff
@@ -0,0 +1,39 @@
+    }
+}
+
+export default function (srcElements: Element[], newParent: Element): Function {
```
No worries either way, but I know on our project we prefer files with a single (default) export, so it's likely I'll be a repeat offender here.

adixon-adobe commented on packages/shared/src/reparent-children.ts:
```diff
@@ -0,0 +1,39 @@
+    placeholderItems: Comment[],
+    srcElements: Element[]
+): void {
+    if (srcElements.length != placeholderItems.length) {
```
I think I can just remove this. This made me realize that we shouldn't collect the children to reparent on firstUpdate, but rather when you open the menu. Otherwise if the hierarchy changes we won't pick it up.

Westbrook commented on packages/split-button/src/SplitButton.ts:
```diff
@@ -158,47 +152,33 @@ export class SplitButton extends SizedMixin(PickerBase) {
         }
-        await this.optionsMenu.updateComplete;
-        if (this.optionsMenu.menuItems.length) {
+        await this.updateComplete;
```
If we can actually move away from the use of `this.optionsMenu.menuItems` (see https://github.com/adobe/spectrum-web-components/pull/1237/files#r586106780 for the logic on why we might not be able to) then we shouldn't get anything useful from the `await` here and can remove.

Westbrook commented on packages/picker/src/Picker.ts:
```diff
@@ -408,12 +413,16 @@ export class PickerBase extends SizedMixin(Focusable) {
-        if (this.optionsMenu.menuItems.length) {
-            this.manageSelection();
+        if (this.open && this.optionsMenu) {
+            await this.optionsMenu.updateComplete;
```
For reference in this example, with the `sp-menu` removed, you'd get `this.menuItems === []` and `this.optionsMenu.menuItems === [MenuItem, MenuItem, MenuItem, MenuItem]`

adixon-adobe commented on packages/picker/src/Picker.ts:
```diff
@@ -421,6 +425,15 @@ export class PickerBase extends SizedMixin(Focusable) {
+        if (!this.open) {
+            this.menuItems = [
+                ...this.querySelectorAll(`sp-menu-item`),
+            ] as MenuItem[];
```
My thinking here is that it should be okay to change the menu items, but that it works best when the menu is closed.

adixon-adobe commented:
It'd be good to update the picker tests so the slotted content tests actually open/select something (and we validate that the slotted content wasn't reparented).

It'd also be good to add a test where we add or remove a menu item and ensure that things still work (and the newly added item is selectable).

I might need some help with those though, and if this otherwise looks good it might make sense to add an issue and merge.

adixon-adobe commented:
I'm seeing some errors in the storybook pages now that I'm digging deeper, so hold on any merge here.

adixon-adobe commented:
@Westbrook I think I've addressed all the feedback you had (except `@query('sp-menu', true)` which doesn't work.

I can add a separate issue for improving some of the tests here. I know the visual diffs caught some things the tests could have caught earlier (but that also means there's at least a little coverage of some scenarios I noticed the tests were missing).

Westbrook reviewed (approved):
LGTM.

Would love to see `reparentChildren` leveraged in `ActiveOverlay` as a stand along PR before we get to the bigger menu changes you'll be looking at next. The reusability of that API is a really great addition to the library! Thanks for the deep work here.
